### PR TITLE
[ROCm] Enable ROCm support for inductor's dynamic_rblock_scaling 

### PR DESCRIPTION
**Pull request by jataylo**
As of ROCm 6.1 [hipDeviceProp_t::regsPerMultiprocessor](https://rocm.docs.amd.com/projects/HIP/en/latest/doxygen/html/structhip_device_prop__t.html#a7390d5b180d63978c81aa971060270b4) is now available allowing us to enable this attribute on ROCm.
```
>>> torch.cuda.get_device_properties(0)
_CudaDeviceProperties(name='AMD Instinct MI250X/MI250', major=9, minor=0, gcnArchName='gfx90a:sramecc+:xnack-', total_memory=65520MB, multi_processor_count=104)
>>> torch.cuda.get_device_properties(0).regs_per_multiprocessor
65536
```

With https://github.com/triton-lang/triton/pull/3962we can extract n_regs and n_spells from a triton binary with AMD backend allowing us to enable inductor's dynamic_rblock_scaling on ROCm initially implemented in https://github.com/pytorch/pytorch/pull/115094

Leaving this in draft until following PRs have landed:
- https://github.com/pytorch/pytorch/pull/129361 to bump the triton commit pin 
- https://github.com/pytorch/pytorch/pull/128449 to allow us to grab warp_size from device properties instead of hard coding 64 on ROCm.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @hongxiayang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/runtime/hints.py
Comment (shunting314) [inline torch/_inductor/runtime/hints.py:115]: nit: 
```
if device_type in ["cuda", "hip"]:
```

**Review by shunting314 — COMMENTED** [review]
Looks good. Can accept after you rebase and make the CI clean.

File: torch/_inductor/runtime/hints.py
Comment (jataylo) [reply torch/_inductor/runtime/hints.py:115]: Thanks, I thought that I tried this way initially but was hitting some dynamo error. I must have gotten it the wrong way around seems to work fine.

There's a bit of cleanup needed in this PR let me re-ping when I'm ready to move from draft

File: torch/_inductor/runtime/triton_heuristics.py
Comment (jataylo) [reply torch/_inductor/runtime/triton_heuristics.py:261]: The device_prop.major >=8 is CUDA specific, so this was a just a simple way to conditionalize this out restriction for ROCm. I will consult if we should have some similar restriction on ROCm side.

**Review by jansel — APPROVED** [review]

**Comment by jataylo** [comment]
Needs a rebase

File: torch/_inductor/runtime/triton_heuristics.py
Comment (jataylo) [reply torch/_inductor/runtime/triton_heuristics.py:261]: Hey @shunting314, ready for review now, we are green any other comments or are we happy to close this?

**Comment by jataylo** [comment]
Hey @shunting314 any more comments or are we good to merge? 

We already have some approval so I'll give it a day or so to see if you have any further comments before merging

**Review by shunting314 — APPROVED** [review]

**Comment by atalman** [comment]
Receive following errors:
```
torch/csrc/cuda/Module.cpp:973:56: error: no member named 'regsPerMultiprocessor' in 'hipDeviceProp_t'
          "regs_per_multiprocessor", &hipDeviceProp_t::regsPerMultiprocessor)
...
torch/csrc/cuda/Module.cpp:988:9: error: variable 'stream' cannot be implicitly captured in a lambda with no capture-default specified
        stream << "_CudaDeviceProperties(name='" << prop.name
        ^
...
torch/csrc/cuda/Module.cpp:986:24: note: lambda expression begins here
      .def("__repr__", [](const hipDeviceProp_t& prop) {
                       ^
``

**Comment by jataylo** [comment]
@atalman @jithunnair-amd @jeffdaily 
Do we know which ROCm version the internal builds are using? regsPerMultiprocessor should have been available since 6.1.0
https://rocm.docs.amd.com/projects/hip-python/en/docs-6.1.0/python_api/hip.html#hip.hip.hipDeviceProp_t.regsPerMultiprocessor

**Comment by jeffdaily** [comment]
> @atalman @jithunnair-amd @jeffdaily Do we know which ROCm version the internal builds are using? regsPerMultiprocessor should have been available since 6.1.0 https://rocm.docs.amd.com/projects/hip-python/en/docs-6.1.0/python_api/hip.html#hip.hip.hipDeviceProp_t.regsPerMultiprocessor

They are using ROCm ~6.0.  Can this code be rocm version ifdef'd somehow?

**Comment by atalman** [comment]
@jeffdaily @jataylo @jithunnair-amd Should we also keep some rocm 6.0 tests in CI to cover for these use cases ?

**Comment by jeffdaily** [comment]
> @jeffdaily @jataylo @jithunnair-amd Should we also keep some rocm 6.0 tests in CI to cover for these use cases ?

A build-only (CPU-only) job would have caught this one, right?  Otherwise we can't afford any more GPU jobs.

**Comment by atalman** [comment]
@jeffdaily this was caught by gpu build job, but having at least cpu only build would be better then nothing

**Comment by jataylo** [comment]
Should be close to merging, final rebase to hopefully get rid of some transient errors.

**Pull request closed without merging**
